File: src/overtake_decision/src/DecisionNode.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <std_msgs/msg/int32.hpp>
#include <string>
#include <algorithm>

class DecisionNode : public rclcpp::Node {
public:
    DecisionNode() : Node("decision_node"), current_lane_(0), consecutive_count_(0), last_parsed_direction_("NONE") {
        // Declare and get parameter
        this->declare_parameter("hysteresis_threshold", 5);
        threshold_ = this->get_parameter("hysteresis_threshold").as_int();

        // Subscribers & Publishers
        vlm_sub_ = this->create_subscription<std_msgs::msg::String>(
            "/vlm_output", 10, std::bind(&DecisionNode::vlm_callback, this, std::placeholders::_1));
        
        lane_pub_ = this->create_publisher<std_msgs::msg::Int32>("/target_lane", 10);

        RCLCPP_INFO(this->get_logger(), "Overtake Decision Node (C++) Started. Waiting for VLM output...");
    }

private:
    void vlm_callback(const std_msgs::msg::String::SharedPtr msg) {
        // 1. Cleaning & Keyword Extraction
        std::string raw_text = msg->data;
        // Convert to uppercase
        std::transform(raw_text.begin(), raw_text.end(), raw_text.begin(), ::toupper);

        std::string clean_direction = "NONE";
        if (raw_text.find("RIGHT") != std::string::npos) {
            clean_direction = "RIGHT";
        } else if (raw_text.find("LEFT") != std::string::npos) {
            clean_direction = "LEFT";
        } else if (raw_text.find("CENTER") != std::string::npos) {
            clean_direction = "CENTER";
        }

        // 2. Hysteresis Filtering
        if (clean_direction == last_parsed_direction_ && clean_direction != "NONE") {
            consecutive_count_++;
        } else {
            consecutive_count_ = 0;
            last_parsed_direction_ = clean_direction;
        }

        // 3. Decision Logic Execution
        if (consecutive_count_ >= threshold_) {
            process_strategy(clean_direction);
        }
    }

    void process_strategy(const std::string& direction) {
        // Strategic mapping
        int target = 0; // Default to racing line (0)
        
        if (direction == "CENTER" || direction == "RIGHT") {
            target = 1; // Overtake via Left Lane
        } else if (direction == "LEFT") {
            target = 2; // Overtake via Right Lane
        }

        if (target != current_lane_) {
            current_lane_ = target;
            std_msgs::msg::Int32 msg;
            msg.data = current_lane_;
            lane_pub_->publish(msg);
            
            std::string lane_name;
            if (target == 0) lane_name = "CENTER";
            else if (target == 1) lane_name = "LEFT";
            else if (target == 2) lane_name = "RIGHT";

            RCLCPP_INFO(this->get_logger(), "STRATEGY UPDATE: Confirmed %s. Switching to %s lane.", direction.c_str(), lane_name.c_str());
        }
    }

    // Members
    int threshold_;
    int current_lane_;
    int consecutive_count_;
    std::string last_parsed_direction_;

    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr vlm_sub_;
    rclcpp::Publisher<std_msgs::msg::Int32>::SharedPtr lane_pub_;
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<DecisionNode>());
    rclcpp::shutdown();
    return 0;
}

File: src/decision_maker/decision_maker/decision_node.py
import rclpy
import json
from rclpy.node import Node
from std_msgs.msg import String, Int32
from rcl_interfaces.srv import SetParameters
from rcl_interfaces.msg import Parameter, ParameterType, ParameterValue

class DecisionNode(Node):
    def __init__(self):
        super().__init__('decision_node')
        # Subscribe to LLM output
        self.subscription = self.create_subscription(
            String,
            'llm_vel',
            self.mode_callback,
            10)
        
        # Publisher for Lane Changing
        self.lane_publisher = self.create_publisher(Int32, '/target_lane', 10)
        
        # Client to set parameters on the controller node
        self.param_client = self.create_client(SetParameters, '/stanley_controller/set_parameters')
        
        self.current_mode = "NORMAL"
        self.get_logger().info('Decision Node started. Waiting for driving mode from LLM...')

        # Subscribe to VLM output for overtaking
        self.vlm_subscription = self.create_subscription(
            String,
            '/vlm_overtake',
            self.vlm_callback,
            10)

    def mode_callback(self, msg):
        try:
            data = json.loads(msg.data)
            raw_text = data.get("action", "").upper()
            self.get_logger().info(f"Received JSON action: {raw_text} (Reason: {data.get('reason', 'N/A')})")
        except json.JSONDecodeError:
            self.get_logger().warn("Received non-JSON input, falling back to raw string")
            raw_text = msg.data.upper()

        detected_mode = None

        # 1. Lane Change Logic
        if "GO_LEFT" in raw_text:
            self.change_lane(1) # 1 is Left
        elif "GO_RIGHT" in raw_text:
            self.change_lane(0) # 0 is Right

        # 2. Driving Mode Logic
        if "DECELERATE" in raw_text:
            detected_mode = "SAFETY"
        elif "ACCELERATE" in raw_text:
            detected_mode = "AGGRESSIVE"
        elif "NORMAL" in raw_text:
            detected_mode = "NORMAL"

        if detected_mode:
            if detected_mode != self.current_mode:
                self.get_logger().info(f'LLM output "{raw_text}" -> Switching to {detected_mode} mode')
                self.current_mode = detected_mode
                self.apply_parameters(detected_mode)
        elif "GO_LEFT" not in raw_text and "GO_RIGHT" not in raw_text:
             # Only warn if it wasn't a lane change command either
            self.get_logger().warn(f'Ignored invalid mode from LLM: "{raw_text}"')

    def vlm_callback(self, msg):
        try:
            data = json.loads(msg.data)
            if not data.get("success", False):
                return
            
            decision = data.get("decision", "").lower()
            self.get_logger().info(f"Received VLM decision: {decision}")
            
            if decision == "center":
                # Car is in front, go left to overtake
                self.change_lane(1) 
            # If decision is "stay", "left", or "right", continue in current lane (do nothing)
            
        except json.JSONDecodeError:
            self.get_logger().error(f"Failed to decode VLM JSON: {msg.data}")

    def change_lane(self, lane_index):
        msg = Int32()
        msg.data = lane_index
        self.lane_publisher.publish(msg)
        lane_name = "LEFT" if lane_index == 1 else "RIGHT"
        self.get_logger().info(f'Lane Change Requested: {lane_name} ({lane_index})')

    def apply_parameters(self, mode):
        req = SetParameters.Request()
        
        # Define parameter sets for each mode
        if mode == "SAFETY":
            # Slow down, lower gains for stability
            req.parameters = [
                Parameter(name='v_scale', value=ParameterValue(type=ParameterType.PARAMETER_DOUBLE, double_value=0.2)),
                Parameter(name='k_e', value=ParameterValue(type=ParameterType.PARAMETER_DOUBLE, double_value=1.0)),
                 Parameter(name='k_h', value=ParameterValue(type=ParameterType.PARAMETER_DOUBLE, double_value=0.5))
            ]
        elif mode == "AGGRESSIVE":
            # Speed up, higher gains for responsiveness
            req.parameters = [
                Parameter(name='v_scale', value=ParameterValue(type=ParameterType.PARAMETER_DOUBLE, double_value=1.0)),
                Parameter(name='k_e', value=ParameterValue(type=ParameterType.PARAMETER_DOUBLE, double_value=1.0)),
                 Parameter(name='k_h', value=ParameterValue(type=ParameterType.PARAMETER_DOUBLE, double_value=0.5))
            ]
        else: # NORMAL
             req.parameters = [
                Parameter(name='v_scale', value=ParameterValue(type=ParameterType.PARAMETER_DOUBLE, double_value=0.5)),
                Parameter(name='k_e', value=ParameterValue(type=ParameterType.PARAMETER_DOUBLE, double_value=1.0)),
                 Parameter(name='k_h', value=ParameterValue(type=ParameterType.PARAMETER_DOUBLE, double_value=0.5))
            ]

        # Call the service asynchronously
        if self.param_client.wait_for_service(timeout_sec=1.0):
            future = self.param_client.call_async(req)
            future.add_done_callback(self.param_response_callback)
        else:
            self.get_logger().warn('Controller parameter service not available!')

    def param_response_callback(self, future):
        try:
            result = future.result()
            self.get_logger().info(f'Parameters updated successfully: {result}')
        except Exception as e:
            self.get_logger().error(f'Failed to update parameters: {e}')

def main(args=None):
    rclpy.init(args=args)
    node = DecisionNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()
